server/src/main/java/io/crate/metadata/bugfix/CorruptedMetadataFixer.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.metadata.bugfix;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.SETTING_CREATION_DATE;
+import static org.elasticsearch.cluster.metadata.IndexMetadata.SETTING_INDEX_UUID;
+
+import java.util.List;
+import java.util.Map;
+
+import javax.annotation.Nonnull;
+
+import org.elasticsearch.cluster.metadata.AliasMetadata;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.settings.Settings;
+
+import com.carrotsearch.hppc.cursors.ObjectObjectCursor;
+
+import io.crate.common.annotations.VisibleForTesting;
+import io.crate.common.collections.Maps;
+import io.crate.execution.ddl.Templates;
+import io.crate.metadata.IndexParts;
+import io.crate.metadata.PartitionName;
+import io.crate.metadata.RelationName;
+
+public class CorruptedMetadataFixer {
```

Comment:
Some high level description of the corrupted state (and how/why it happened) would be nice.
For example:

```suggestion

/**
 * Class to fix metadata corruption caused by SWAP TABLE statements like
 *
 * <pre>
 *  ALTER CLUSTER SWAP TABLE "telegraf"."metrics_new" TO "telegraf.metrics";
 * </pre>
 *
 * <p>
 * Quoting the full name ("telegraf.metrics") instead of schema and table
 * separate ("telegraf"."metrics") corrupted the metadata. It assumed
 * "telegraf.metrics" is the table name, and used "doc" (or current schema from
 * search path) as schema. A `.` is used in the template and index name to encode schema, table
 * and partition name. Having a `.` within the table name part broke the
 * encoding.
 *
 * <h2>Valid encoding schemes</h2>
 *
 * <ul>
 *  <li>{@code <table_name>}</li>
 *  <li>{@code <schema>.<table_name>}</li>
 *  <li>{@code .partitioned.<table_name>.[<ident>]}</li>
 *  <li>{@code <schema>..partitioned.<table_name>.[<ident>]}</li>
 * </ul>
 *
 * With "metrics" as table name:
 *
 * <ul>
 *  <li>{@code metrics}</li>
 *  <li>{@code custom_schema.metrics}</li>
 *  <li>{@code .partitioned.metrics.[<ident>]}</li>
 *  <li>{@code custom_schema..partitioned.metrics.[<ident>]}</li>
 * </ul>
 *
 * With the illegal `.` in the table name:
 *
 * <ul>
 *  <li>{@code tbl.with_dots} → incidentally did the right thing, moved table to correct schema</li>
 *  <li>{@code custom_schema.tbl.with_dots} → SWAP TABLE validations already prevented this; Would cause exception in {@link IndexParts}</li>
 *  <li>{@code .partitioned.tbl.with_dots.[<ident>]} → 5 parts, causes exception in {@link IndexParts} because second part is "partitioned" instead of empty
 *  <li>{@code custom_schema..partitioned.tbl.with_dots.[<ident>]} → SWAP TABLE validations prevented this, 6 parts would cause exceptions in {@link IndexParts}</li>
 * </ul>
 *
 *
 * <p>
 * See https://github.com/crate/crate/issues/13380
 * </p>
 **/
public class CorruptedMetadataFixer {
```


Maybe even extended with the x(?) scenarios:

- partitioned "schema"."table" to partitioned "schema.table"
- partitioned "schema.table" to partitioned "schema"."table"
- partitioned "schema.table" to partitioned "schema.table"
- partitioned "schema"."table" to non-partitioned "schema.table"
- partitioned "schema.table" to non-partitioned "schema"."table"
- partitioned "schema.table" to non-partitioned "schema.table"
- non-partitioned "schema"."table" to partitioned "schema.table"
- non-partitioned "schema.table" to partitioned "schema"."table"
- non-partitioned "schema.table" to partitioned "schema.table"

Not necessary as it appears as if the costs cover it, but maybe some short elaboration on how the `partition_by` meta hiccup and the loss of data happened would be good